cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
 	SkipPhases []string
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UpgradeConfiguration contains a list of elements that is specific "kubeadm upgrade"-only runtime information.
+type UpgradeConfiguration struct {
+	metav1.TypeMeta
+
+	// ApiServerManifestPath is the path to API server manifest
+	// Default: /etc/kubernetes/manifests/kube-apiserver.yaml
+	ApiServerManifestPath string
```

Comment:
Just a question? Do we need to consider the case of dry-run? 

Looks like to me `kubeadm upgrade diff` (only this CLI needs this flag I think)  is not created for the comparison with dry-run manifest.

https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/cmd/kubeadm/app/cmd/init.go#L429-L431

Reply:
i still haven't checked the source code about where these custom manifest paths are used (currently on phone).

if the use case is important this could have been a custom folder with standard manifest file names. not custom manifest path for each component.

IMO, we should deprecate the flags and not add these options in config.

Reply:
update: I paste the wrong link, it should be, 

https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/cmd/kubeadm/app/cmd/init.go#L457-L462

So, it will use the hard-coded path to write the manifest, the same for join and reset, but if it's dry-run mode, the path can be a little different and can be set by the env `KUBEADM_INIT_DRYRUN_DIR`. 

But I think the dir is just a temp dir and my guess for the `kubeadm upgrade diff` is not created for `dry-run`, I am fine to remove it from config file.

@SataQiu @pacoxu  any inputs for this?

Reply:
I removed them from the upgrade config API type.